### ISSUE TITLE
Sync Batch Norm for MXNet

### ISSUE THREAD

**Issue by jasonliu747** [horovod/horovod#2087]
There are sync batch norm implementations for PyTorch and Tensorflow in Horovod. Is MXNet in our TODO list?

**Comment by tgaddair** [comment]
Hey @jasonliu747, @romerojosh might have some thoughts on that.  If you're interested in putting together a PR, we'd definitely appreciate it!

**Comment by jasonliu747** [comment]
Cool, Thanks! @tgaddair 
@romerojosh It would be appreciated if you could share your current progress and thoughts!

**Comment by romerojosh** [comment]
Hi @jasonliu747, I haven't currently done any work on adding a sync BN implementation for MXNet. The tricky part right now is that the Horovod MXNet implementation doesn't have any examples showing how to integrate Horovod operations into the forward/backwards pass. For example, the PyTorch Horovod allreduce has an autograd function version that enables it to be used within the compute graph defined here:
https://github.com/horovod/horovod/blob/3e4541881bc659ec4ff59383df917aa319be7cae/horovod/torch/mpi_ops.py#L144-L156

MXNet appears to have similar functionality described here: https://mxnet.apache.org/versions/1.2.1/api/python/autograd/autograd.html

If the MXNet autograd functions work similarly those in PyTorch, adding a Sync BN for MXNet _might_ be as simple as:
1. defining these autograd functions for the primitive Horovod ops (allreduce, bcast, allgather) in MXNet as is done in PyTorch (see torch/mpi_ops.py)
2. Porting the PyTorch Sync BN implementation (torch/sync_batch_norm.py) to MXNet.

**Comment by shuo-ouyang** [comment]
Hi @romerojosh , thanks for your advice. I implement SyncBN for MXNet by registering Horovod allreduce as a mxnet custom operator, and write forward and backward methods for this allreduce op. Then I create SyncBN class, inherit it from `gluon.nn.Block`, and overload the forward function. In my experiments, this approach works well for gluon interface. However, due to the static property of symbolic programming, it does not work for symbol interface.

**Comment by romerojosh** [comment]
Hi @shuo-ouyang, this sounds great! Any interest in posting a PR with the implementation?

I think it is ok that it only works with gluon. Symbolic seems to be slowly being deprecated in favor of gluon anyway.

**Comment by shuo-ouyang** [comment]
Hi @romerojosh, sorry for delay response. Our current SyncBN implementation needs to re-write BN operator manually, and suffers from some performance issues.
